src/lib.rs

```diff
@@ -251,6 +256,133 @@ impl DstLayout {
             _trailing_slice_elem_size: Some(mem::size_of::<T>()),
         }
     }
+
+    /// TODO
+    ///
+    /// The caller is responsible for ensuring that `addr + bytes_len` does not
+    /// overflow `usize`.
+    ///
+    /// # Panics
+    ///
+    /// If `addr + bytes_len` overflows `usize`, `validate_cast` may panic, or
+    /// it may return incorrect results. No guarantees are made about when
+    /// `validate_cast` will panic. The caller should not rely on
+    /// `validate_cast` panicking in any particular condition, even if
+    /// `debug_assertions` are enabled.
+    const fn _validate_cast(
+        &self,
+        addr: usize,
+        bytes_len: usize,
+        cast_type: _CastType,
+    ) -> Option<(usize, usize, usize)> {
+        let base_size = self._base_layout.size();
+
+        // LEMMA 0: max_slice_bytes + base_size == bytes_len
+        //
+        // LEMMA 1: base_size <= bytes_len:
+        // - If `base_size > bytes_len`, `bytes_len.checked_sub(base_size)`
+        //   returns `None`, and we return.
+        //
+        // TODO(#67): Once our MSRV is 1.65, use let-else:
+        // https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements
+        let max_slice_bytes = if let Some(max_byte_slice) = bytes_len.checked_sub(base_size) {
+            max_byte_slice
+        } else {
+            return None;
+        };
+
+        let (elems, self_bytes) = if let Some(elem_size) = self._trailing_slice_elem_size {
+            // TODO(#67): Once our MSRV is 1.65, use let-else:
+            // https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements
+            let elem_size = if let Some(elem_size) = NonZeroUsize::new(elem_size) {
+                elem_size
+            } else {
+                panic!("attempted to cast to slice type with zero-sized element");
```

Comment:
Why the panic here? Is the caller also responsible for ensuring that `_trailing_slice_elem_size` is non-zero?

Reply:
Yeah, good point - I'll need to add that to the doc comment. The idea is that the calling code (in zerocopy) just gets to do `T::LAYOUT.validate_cast(...)`, and it is guaranteed to panic if the user (outside of zerocopy) provided an unsupported type. This is consistent with the documented behavior on the various `Ref` constructors.

I have https://github.com/google/zerocopy/issues/325 filed to track an alternate approach at some point in the future.

Reply:
If ZST slices are generally unsupported, should `DstLayout.trailing_slice_elem_size` have type `Option<NonZeroUsize>` instead?

Reply:
In theory I like the idea of being able to have code like this fail at compile time. However, what do you do about a user who wants to write code like this?

```rust
pub fn do_thing<T: ?Sized + KnownLayout>() {
    if T::LAYOUT.trailing_slice_elem_type == Some(0) {
        // Do workaround thing
    }
    // Do other thing
}
```

As in, you can't say that it's a compile-time error to have code that _could_ trigger this panic because the user might have their own logic that ensures it'll never be encountered at runtime.

Reply:
Just wanted to call out that I moved this logic to the top of the method so that it unconditionally panics. Previously, if certain runtime errors were encountered first, the method would return `None` before encountering the 0-length-slice-element and panicking.